### PR TITLE
Add lock to config reload/load_minigraph

### PR DESCRIPTION
**Pull request by lolyu**
#### What I did
In some cases, if multiple config reload/load_minigraph are running in parallel, they might leave the system in an error state.
In this PR, a flock is added to config reload/load_minigraph so they will not run in parallel.
The file lock is binding to `/etc/sonic/reload.lock`.

This is to fix issue: [#19855](https://github.com/sonic-net/sonic-buildimage/issues/19855)

* Microsoft ADO (number only): 28877643

Signed-off-by: Longxiang Lyu <lolv@microsoft.com>

#### How I did it
Add flock utility and decoate the `reload` and `load_minigraph` with the `try_lock` to ensure the lock is acquired before reload/load_minigraph.

#### How to verify it
UT and on testbed.

#### New command output (if the output of a command-line utility has changed)
* reload with locking success
```
# config reload
Acquired lock on /etc/sonic/reload.lock
Clear current config and reload config in config_db format from the default config file(s) ? [y/N]: y
Disabling container monitoring ...
Stopping SONiC target ...
Running command: /usr/local/bin/sonic-cfggen -j /etc/sonic/init_cfg.json -j /etc/sonic/config_db.json --write-to-db
Running command: /usr/local/bin/db_migrator.py -o migrate
Running command: /usr/local/bin/sonic-cfggen -d -y /etc/sonic/sonic_version.yml -t /usr/share/sonic/templates/sonic-environment.j2,/etc/sonic/sonic-environment
Restarting SONiC target ...
Enabling container monitoring ...
Reloading Monit configuration ...
Reinitializing monit daemon
Released lock on /etc/sonic/reload.lock
```
* reload with locking failure
```
# config reload
Failed to acquire lock on /etc/sonic/reload.lock
```

### CODE REVIEW

**Comment by bingwang-ms** [comment]
You may need to check all the reference of `reload` and `load_minigraph` as the signature is change. I remember the functions are called in some other scripts.
Another option is to add a default value for the new argument `bypass_lock`.

File: config/main.py [hunk 1]
```diff
@@ -1753,9 +1760,11 @@ def list_checkpoints(ctx, verbose):
 @click.option('-n', '--no_service_restart', default=False, is_flag=True, help='Do not restart docker services')
 @click.option('-f', '--force', default=False, is_flag=True, help='Force config reload without system checks')
 @click.option('-t', '--file_format', default='config_db',type=click.Choice(['config_yang', 'config_db']),show_default=True,help='specify the file format')
+@click.option('-b', '--bypass-lock', default=False, is_flag=True, help='Do reload without acquiring lock')
 @click.argument('filename', required=False)
 @clicommon.pass_db
-def reload(db, filename, yes, load_sysinfo, no_service_restart, force, file_format):
+@try_lock(SYSTEM_RELOAD_LOCK, timeout=0)
+def reload(db, filename, yes, load_sysinfo, no_service_restart, force, file_format, bypass_lock):
     """Clear current configuration and import a previous saved config DB dump file.
```
Comment (bingwang-ms) [inline config/main.py:1768]: If some error happened during reload the the process aborted, will the lock be released?

Comment (lolyu) [reply config/main.py:1768]: the lock is automatically dropped when the file is closed, when the process is aborted, the system will cleanup the file opened.

**Review by bingwang-ms — APPROVED** [review]

**Review by zjswhhh — APPROVED** [review]
LGTM.

Should we cherry-pick this into other release branches?

**Comment by lolyu** [comment]
Hi @bingwang-ms @zjswhhh, added another commit to improve the log, please help review, thanks!

**Review by zjswhhh — APPROVED** [review]
lgtm.

**Pull request merged**